File: leetcode/1001-1999/1004-max-consecutive-ones-iii/MaxConsecutiveOnesIII.ts
function longestOnes(numbers: number[], k: number): number {
  let right = 0;
  let left = 0
  let length = numbers.length;

  while (right < length) {
    k -= 1 - numbers[right++];

    if (k < 0) {
      k += 1 - numbers[left];
      left++;
    }
  }

  return right - left;
};

console.log(longestOnes([1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0], 2))

File: leetcode/1001-2000/1001-1100/1004-max-consecutive-ones-iii/MaxConsecutiveOnesIII.java
public class MaxConsecutiveOnesIII {
  public static void main(String[] args) {
    int[] numbers = new int[] { 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0 };

    int k = 2;

    System.out.println(new Solution().longestOnes(numbers, k));
  }
}

class Solution {
  public int longestOnes(int[] numbers, int k) {
    int left = 0;
    int right = 0;
    int max = 0;
    int length = numbers.length;

    while (right < length) {
      if (numbers[right] == 0) {
        k--;
      }

      right++;

      while (k < 0) {
        if (numbers[left] == 0) {
          k++;
        }

        left++;
      }

      max = Math.max(max, right - left);
    }

    return max;
  }
}
